### DISCUSSION TITLE
How to check autoDispose provider is exist?

### DISCUSSION THREAD

**Discussion by td026544** [rrousselGit/riverpod#781] [Q&A] [answered]
How can I check provider instance is exist?

**Comment by rrousselGit** [comment]
Thee is no such thing. Why do you need this?

**Reply by td026544** [reply]
I use MQTT to build chat room app,

If user  receive message and in chat room page , I want to add message in the provider.
```
 if(read(chatMessagesStateNotifierProvider).exist){
      read(chatMessagesStateNotifierProvider(chatId).notifier).addMessage();
  }
```

**Comment by matuella** [comment]
I also needed this because my provider perform side-effects when it calls its `_create`, so I would like to know if it exists before actually reading/watching/listening them. I can exemplify with a `flutter use-case`:

I have a widget that consumes an `autoDisposable` provider which must be reloaded once some particular operation occurs. This means that I would like to `read` that `autoDisposable` provider and perform the operation ONLY if it's alive, because it makes no sense to reload something that it's not even instantiated. Yes, I could workaround this by making all of the `_create` functions pure (meaning, no side-effects), but I don't think this is a rare use-case and could be supported out of the box by `riverpod`. At least for me, I've found myself working around this particular use-case a couple of times.

**Reply by rrousselGit** [reply]
Why is it that the code which wanted to read the provider can do without reading it?

**Reply by matuella** [reply]
Sorry, I don't quite get your question, but to be more specific: the provider in question is a `StateNotifierProvider` which holds a bunch of logic and has multiple states. When this particular provider is alive (meaning, it has listeners), it must be read (`ref.read(provider.notifier).reload()`, and reload being this custom function that I've created) - but only if alive.

It'll behave properly if the provider is alive and has listeners, but the problem is when it isn't, because it'll create the provider (when calling `ref.read(provider.notifier)`, call that `reload` function, and then dispose it. Having something like `ref.isAlive(provider)` would allow me to avoid that.

**Reply by rrousselGit** [reply]
Are you looking for `ref.refresh(provider)`?

**Comment by rrousselGit** [comment, marked as answer]
With 2.1.0, you will now be able to do:

```dart
ref.exists(provider);
container.exists(provider);
widgetRef.exists(provider);
```
